### PR TITLE
volatile chapter

### PR DESCRIPTION
**Pull request by Lokathor**
This chapter adds an introduction to "volatile" semantics, which are a keystone to all MMIO usage, including MMIO usage on the GBA.

### CODE REVIEW

**Comment by Lokathor** [comment]
@RalfJung I know that you don't do GBA stuff specifically, but since this chapter has to do with general unsafe/volatile understanding would you mind giving it a read? It's not too long.

File: book_src/volatile.md [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+  pub fn read(self) -> T {
+    unsafe { self.nn.as_ptr().read_volatile() }
+  }
+  pub fn write(self, t: T) {
+    unsafe { self.nn.as_ptr().write_volatile(t) }
+  }
+}
+```
+
+Cool! Abstractions are fun.
```
Comment (RalfJung) [inline book_src/volatile.md:196]: This abstraction should require `T: Copy` though, otherwise one can wreck all sorts of havoc. ;)

**Review by RalfJung — COMMENTED** [review]
This is great! I really enjoyed reading it. A lot of it could be put straight into the Nomicon, actually, which IMO would be a good idea.

One aspect you do not touch is "references to MMIO memory are bad because the compiler may introduced extra reads/writes". I am not sure of the broader context here; that is something that I think needs to be mentioned in a general "Rust MMIO introduction" but maybe is not needed here.

File: book_src/volatile.md (hunk 1 above)
Comment (Lokathor) [reply book_src/volatile.md:196]: Probably yeah, for good measure. Though that's somewhat covered by "Must be a valid MMIO location for `T`"  I guess if someone has mmio with non-copy types they're in such a weird place they really need to reconsider things from scratch and use a different type.

**Comment by Lokathor** [comment]
Ah, that is an important extra detail. I'll add it in sometime this weekend.

**Comment by Lokathor** [comment]
@RalfJung The Nomicon feels "understaffed", but feel free to use some/all of this chapter text in any eventual MMIO portion you add to it.

**Pull request merged**